File: sorting/inversions.cpp
#include <iostream> 

using namespace std;


void merge(int * a, int size, int med, long* inv){
    //копируем first part of массив a
    int* tmp_arr = new int[med];
    for (int i=0; i<med; ++i){
        tmp_arr[i] = a[i];
    }
    int i = 0;
    int j = 0;
    while (i + j < size){
        if ((j == size - med) || ((i < med) && (tmp_arr[i] <= a[med+j]))){
            a[i + j] = tmp_arr[i];
            ++i;
        }
        else{
            if (j < size-med){
                a[i + j] = a[med+j];
                *inv += med-i;
                ++j;
            }
        }
    }
}


void merge_sort(int* a, int size, long* inv){
    if (size > 1){
        int med = size / 2;
        merge_sort(a, med, inv);
        merge_sort(a+med, size-med, inv);
        merge(a, size, med, inv);
    }
}


int main()
{
    int n = 0;
    int * arr;
    long inv = 0;
    cin >> n;
    arr = new int[n];
    for(int i=0; i<n; ++i) cin >> arr[i];
    merge_sort(arr, n, &inv);
    cout << inv;
}

File: sorting/inversions.py
#!/usr/bin/python3
def merge(a, b, inv):
    n = len(a)
    m = len(b)
    i = 0
    j = 0
    c = []
    while (i + j) < (n + m):
        if ((j == m) or ((i < n) and (a[i] <= b[j]))):
            c.append(a[i])
            i += 1
        else:
            c.append(b[j])
            inv += n - i
            j += 1
    return c, inv


def merge_sort(a, inv):
    n = len(a)
    if n > 1:
        l = a[:n // 2]
        r = a[n // 2:]
        l, inv = merge_sort(l, inv)
        r, inv = merge_sort(r, inv)
        a, inv = merge(l, r, inv)
    return a, inv


if __name__ == "__main__":
    n = int(input())
    line = input()
    a = list(map(int, line.split()))
    inv = 0
    a, inv = merge_sort(a, inv)
    print(inv)